test/e2e/volume_provisioning.go

```diff
@@ -125,6 +126,56 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
 
 			testDynamicProvisioning(c, claim)
 		})
+
+		It("should test that deleting a claim before the volume is provisioned deletes the volume. [Volume]", func() {
+			// This case tests for the regressions of a bug fixed by PR #21268
+			// REGRESSION: Deleting the PVC before the PV is provisioned can result in PV
+			// not being deleted
+
+			// Given the small race window, attempt to produce it iteratively
+			var raceAttempts int = 100
+
+			By("creating a StorageClass")
+			class := newStorageClass()
+			class.Name = ns + "-sc" // pseudo-random Storage Class name
+			class, err := c.Storage().StorageClasses().Create(class)
+			defer c.Storage().StorageClasses().Delete(class.Name, nil)
+			Expect(err).NotTo(HaveOccurred())
+
+			By(fmt.Sprintf("Creating and deleting PersistentVolumeClaims for %d iterations", raceAttempts))
+			claim := newClaim(ns, false)
+			claim.Annotations[storageutil.StorageClassAnnotation] = class.Name
+			for i := 0; i < raceAttempts; i++ {
+				tmpClaim, err := c.Core().PersistentVolumeClaims(ns).Create(claim)
```

Comment:
The issue is that claims will be created with a random name on the `.Create()` call which we need to capture for the `.Delete()` call.  Also, if there is any latency in the delete call and the PVCs are using a predefined name, it's possible to cause a creating PVC to collide with a PVC waiting for deletion.